Source/JavaScriptCore/runtime/MicrotaskQueue.cpp

```diff
@@ -37,22 +41,106 @@ namespace JSC {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(MicrotaskQueue);
 
-void QueuedTask::run()
+QueuedTask::Status QueuedTask::tryRun()
 {
+    if (m_dispatcher)
+        return m_dispatcher->run(*this);
+
     if (!m_job.isObject())
-        return;
+        return Status::Executed;
+
     JSObject* job = jsCast<JSObject*>(m_job);
-    JSGlobalObject* globalObject = job->globalObject();
-    runJSMicrotask(globalObject, m_identifier, job, m_arguments[0], m_arguments[1], m_arguments[2], m_arguments[3]);
+    runJSMicrotask(m_globalObject, m_identifier, job, m_arguments[0], m_arguments[1], m_arguments[2], m_arguments[3]);
+    return Status::Executed;
+}
+
+bool QueuedTask::isRunnable()
+{
+    if (!m_dispatcher)
+        return true;
+    return m_dispatcher->isRunnable();
```

Comment:
We should store `m_dispatcher` in a local `RefPtr` before calling `isRunnable()`.